File: tauri-app/src/ipc/midi.ts
// Typed wrappers for the MIDI input IPC commands and events.
//
// Source of truth on the Rust side:
//   tauri-app/src-tauri/src/midi.rs  (midi_status, mock_key commands)
//   crates/midi/src/mock.rs::KEY_MAP (key → MIDI-note mapping)

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

// ── Types ─────────────────────────────────────────────────────────────────────

/** Status of the active MIDI input source. */
export interface MidiStatus {
  /** `"live"` when a real USB-MIDI device is connected, `"mock"` otherwise. */
  kind: "live" | "mock";
  /** Connected port name when `kind === "live"`, `null` otherwise. */
  port: string | null;
}

/**
 * A serialised MIDI event payload from the backend `midi_event` push event.
 *
 * Mirror of `tauri-app/src-tauri/src/midi.rs::NoteEventPayload`.
 * The backend projects `NoteEvent` into this flat, JSON-friendly shape so
 * the frontend never has to handle serde enum tags.
 */
export interface NoteEventPayload {
  /** MIDI note number (0–127). */
  note: number;
  /** `true` = note-on, `false` = note-off. */
  on: boolean;
  /** Velocity (0–127); 0 for note-offs. */
  velocity: number;
  /** Engine timestamp in microseconds (preserved from the source). */
  timestamp_us: number;
}

// ── Key map ───────────────────────────────────────────────────────────────────

/**
 * Computer-key → MIDI-note mapping for the mock keyboard.
 *
 * SOURCE OF TRUTH: `crates/midi/src/mock.rs::KEY_MAP`.
 * The number row plays a C-major (white-key) scale from middle C (60):
 *
 * ```
 *   1   2   3   4   5   6   7   8   9   0
 *   C4  D4  E4  F4  G4  A4  B4  C5  D5  E5
 *   60  62  64  65  67  69  71  72  74  76
 * ```
 *
 * Keep this in sync with the Rust `KEY_MAP` constant when that changes.
 */
export const MOCK_KEY_MAP: ReadonlyArray<[string, number]> = [
  ["1", 60], // C4
  ["2", 62], // D4
  ["3", 64], // E4
  ["4", 65], // F4
  ["5", 67], // G4
  ["6", 69], // A4
  ["7", 71], // B4
  ["8", 72], // C5
  ["9", 74], // D5
  ["0", 76], // E5
];

/** The set of keys that map to MIDI notes (for quick membership checks). */
export const MOCK_KEY_SET: ReadonlySet<string> = new Set(
  MOCK_KEY_MAP.map(([k]) => k),
);

// ── IPC helpers ───────────────────────────────────────────────────────────────

/** Event name the backend emits serialised `NoteEvent`s on. */
const EVENT_MIDI = "midi_event";

/**
 * Subscribe to `midi_event` backend events.
 *
 * Returns the Tauri unlisten function; call it in `onCleanup`.
 *
 * **Debugging snippet** (paste in the webview console):
 * ```js
 * const { listen } = window.__TAURI__.event;
 * listen('midi_event', e => console.log('midi', e.payload));
 * ```
 */
export function onMidiEvent(
  cb: (ev: NoteEventPayload) => void,
): Promise<UnlistenFn> {
  return listen<NoteEventPayload>(EVENT_MIDI, (e) => cb(e.payload));
}

/** Fetch the current MIDI input status (kind + port name). */
export function midiStatus(): Promise<MidiStatus> {
  return invoke<MidiStatus>("midi_status");
}

/**
 * Simulate a computer-key press or release on the mock keyboard.
 *
 * `down=true` triggers a note-on; the backend auto-enqueues a note-off
 * after the sustain window, so `down=false` is accepted but is a no-op.
 *
 * Rejects when a live device is active (mock keys must not inject fake events).
 *
 * Only call this when `midiStatus().kind === "mock"`.
 */
export function mockKey(key: string, down: boolean): Promise<void> {
  // The Rust side takes a `char`; single-character strings map cleanly.
  const ch = key.length === 1 ? key : key[0];
  return invoke<void>("mock_key", { key: ch, down });
}

File: tauri-app/src-tauri/src/midi.rs
//! MIDI input service for the Tauri backend.
//!
//! Tries to open a live USB-MIDI device on startup; falls back to a
//! [`MockKeyboard`] so every screen is drivable in a sandbox with no hardware.
//! Events are drained each tick by the transport thread in `lib.rs`, which
//! (1) emits a `midi_event` webview event and (2) feeds each event into
//! `Composer::ingest` so StepRecord / LiveRecord work identically to the TUI.
//!
//! CLAUDE.md invariant: we never block the MIDI callback path. `LiveInput`
//! buffers events in a channel; we just drain it from the tick thread.

use std::sync::Mutex;

use rockcraft_core::{Composer, NoteEvent, NoteEventKind};
use rockcraft_midi::{LiveInput, MockKeyboard, NoteSource};
use serde::Serialize;

// ── InputSource ───────────────────────────────────────────────────────────────

/// The active event source.
pub enum InputSource {
    /// A live USB-MIDI device (midir-backed).
    Live(LiveInput),
    /// Computer-keyboard mock — present when no hardware port was found.
    Mock(MockKeyboard),
    /// Placeholder while the service is being initialised (never stored long).
    None,
}

impl InputSource {
    /// Drain pending events from whatever source is active.
    pub fn drain_events(&mut self) -> Vec<NoteEvent> {
        match self {
            InputSource::Live(l) => l.events(),
            InputSource::Mock(m) => m.events(),
            InputSource::None => Vec::new(),
        }
    }

    /// Human-readable kind string for the status command.
    pub fn kind_str(&self) -> &'static str {
        match self {
            InputSource::Live(_) => "live",
            InputSource::Mock(_) => "mock",
            InputSource::None => "none",
        }
    }

    /// Connected port name, if any.
    pub fn port_name(&self) -> Option<String> {
        match self {
            InputSource::Live(l) => Some(l.port_name().to_string()),
            _ => None,
        }
    }
}

// ── MidiState ─────────────────────────────────────────────────────────────────

/// Tauri-managed MIDI service state.
pub struct MidiState {
    /// The active input source, behind a mutex so the tick thread and commands
    /// can safely share it.
    pub source: Mutex<InputSource>,
}

impl MidiState {
    /// Initialise: try a live connection first; fall back to mock.
    pub fn new() -> Self {
        let source = match LiveInput::connect("") {
            Ok(live) => InputSource::Live(live),
            Err(_) => InputSource::Mock(MockKeyboard::new()),
        };
        Self {
            source: Mutex::new(source),
        }
    }
}

impl Default for MidiState {
    fn default() -> Self {
        Self::new()
    }
}

// ── IPC types ─────────────────────────────────────────────────────────────────

/// Serialisable status payload returned by `midi_status`.
#[derive(Debug, Clone, Serialize)]
pub struct MidiStatus {
    /// `"live"` or `"mock"`.
    pub kind: String,
    /// Connected port name when `kind == "live"`, otherwise `null`.
    pub port: Option<String>,
}

/// Serialisable payload for the `midi_event` webview event.
///
/// `NoteEvent` itself does not derive `Serialize` (it's a pure `core` type),
/// so we project the fields we care about into this flat, JSON-friendly form.
#[derive(Debug, Clone, Serialize)]
pub struct NoteEventPayload {
    /// MIDI note number (0–127).
    pub note: u8,
    /// `true` = note-on, `false` = note-off.
    pub on: bool,
    /// Velocity (0–127); 0 when this is a note-off.
    pub velocity: u8,
    /// Engine timestamp in microseconds.
    pub timestamp_us: u64,
}

impl From<NoteEvent> for NoteEventPayload {
    fn from(ev: NoteEvent) -> Self {
        let (on, velocity) = match ev.kind {
            NoteEventKind::On { velocity } => (true, velocity.value()),
            NoteEventKind::Off => (false, 0),
        };
        Self {
            note: ev.note.value(),
            on,
            velocity,
            timestamp_us: ev.timestamp_us,
        }
    }
}

// ── Command helpers (free functions for headless testing) ─────────────────────

/// Return the current MIDI input status.
pub fn midi_status(state: &MidiState) -> MidiStatus {
    let source = state.source.lock().expect("midi source mutex poisoned");
    MidiStatus {
        kind: source.kind_str().to_string(),
        port: source.port_name(),
    }
}

/// Simulate a computer-key press/release on the mock keyboard.
///
/// Returns `Err` when the active source is live (we must not inject fake events
/// while a real piano is connected).
pub fn mock_key(state: &MidiState, key: char, down: bool) -> Result<(), String> {
    let mut source = state.source.lock().expect("midi source mutex poisoned");
    match &mut *source {
        InputSource::Mock(kb) => {
            if down {
                kb.press(key);
            }
            // Note-off is handled automatically: MockKeyboard enqueues its own
            // note-off SUSTAIN_MS after the press, so we only need to act on
            // key-down here.  Key-up is a no-op (the note-off is already queued).
            Ok(())
        }
        InputSource::Live(_) | InputSource::None => {
            Err("mock_key is not available while a live MIDI device is connected".to_string())
        }
    }
}

/// Drain pending events, feed each one into the composer, and return
/// serialisable payloads (for `midi_event` emission) plus accumulated effects.
pub fn drain_and_ingest(
    midi: &MidiState,
    composer: &mut Composer,
) -> (Vec<NoteEventPayload>, Vec<rockcraft_core::Effect>) {
    let events = midi
        .source
        .lock()
        .expect("midi source mutex poisoned")
        .drain_events();
    let mut all_effects = Vec::new();
    let mut payloads = Vec::with_capacity(events.len());
    for ev in events {
        let effects = composer.ingest(ev);
        all_effects.extend(effects);
        payloads.push(NoteEventPayload::from(ev));
    }
    (payloads, all_effects)
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use rockcraft_core::{MidiNote, NoteEventKind, Velocity};

    fn make_mock_state() -> MidiState {
        MidiState {
            source: Mutex::new(InputSource::Mock(MockKeyboard::new())),
        }
    }

    #[test]
    fn mock_key_press_enqueues_note() {
        let state = make_mock_state();
        // Press '1' (maps to C4 = 60)
        mock_key(&state, '1', true).expect("mock_key should succeed on mock source");
        // Give a tiny sleep so the event timestamp passes
        std::thread::sleep(std::time::Duration::from_millis(200));
        // Drain should yield at least the note-on
        let mut src = state.source.lock().unwrap();
        let events = src.drain_events();
        assert!(
            !events.is_empty(),
            "expected note events after mock_key press"
        );
        let note_on = events
            .iter()
            .find(|e| matches!(e.kind, NoteEventKind::On { .. }));
        assert!(note_on.is_some(), "expected a note-on event");
        assert_eq!(note_on.unwrap().note.value(), 60, "C4 should be midi 60");
    }

    #[test]
    fn mock_key_on_live_source_returns_err() {
        // We cannot open a real MIDI device in CI, so simulate a Live source
        // by checking the error path through the mock; instead we just verify
        // the None variant also returns Err.
        let state = MidiState {
            source: Mutex::new(InputSource::None),
        };
        let result = mock_key(&state, '1', true);
        assert!(
            result.is_err(),
            "mock_key on None/Live source must return Err"
        );
    }

    #[test]
    fn midi_status_mock_returns_mock_kind() {
        let state = make_mock_state();
        let status = midi_status(&state);
        assert_eq!(status.kind, "mock");
        assert!(status.port.is_none());
    }

    #[test]
    fn drain_and_ingest_with_scripted_source() {
        use rockcraft_midi::ScriptedSource;

        // Build a ScriptedSource with one note-on at t=0 and note-off at t=500_000.
        let note = MidiNote::new(60).unwrap();
        let vel = Velocity::new(80).unwrap();
        let events_in = vec![NoteEvent::on(note, vel, 0), NoteEvent::off(note, 500_000)];
        let mut src = ScriptedSource::new(events_in);
        // Advance so both events are due.
        src.advance(600_000);

        let state = MidiState {
            source: Mutex::new(InputSource::Mock(MockKeyboard::new())),
        };
        // Replace mock source with scripted by draining and feeding manually.
        // (ScriptedSource doesn't implement the same Mutex wrapper, but we can
        // test drain_and_ingest indirectly via the mock after pressing a key.)
        //
        // Test the composer ingest path directly:
        let mut composer = Composer::new();
        // Switch to StepRecord so ingest places notes.
        composer
            .apply(
                rockcraft_core::action_from_name("toggle_record_arm", &serde_json::json!({}))
                    .unwrap(),
            )
            .unwrap();

        // Feed events manually through the NoteSource trait.
        use rockcraft_midi::NoteSource as _;
        let drained: Vec<NoteEvent> = src.events();
        assert_eq!(drained.len(), 2, "expected 2 events from scripted source");

        // Verify timestamp_us is preserved.
        assert_eq!(drained[0].timestamp_us, 0);
        assert_eq!(drained[1].timestamp_us, 500_000);

        // Feed into composer ingest — in StepRecord the note-on should place a note.
        let mut placed_effects = Vec::new();
        for ev in &drained {
            placed_effects.extend(composer.ingest(*ev));
        }
        let snap = composer.snapshot();
        assert!(
            !snap.notes.is_empty(),
            "StepRecord ingest of note-on should place a note"
        );
    }
}
